### ISSUE TITLE
Reports API

### ISSUE THREAD

**Issue by sreeise** [sreeise/graph-rs#231]
Meta bug for Reports API. Reference: https://docs.microsoft.com/en-us/graph/api/resources/report?view=graph-rest-1.0

- [ ] Microsoft Teams device usage #232
- [ ] Microsoft Teams user activity #232
- [ ] Outlook activity #233
- [ ] Outlook app usage #233
- [ ] Outlook mailbox usage #233
- [ ] Office 365 activiations #234
- [ ] Office 365 active usres #234
- [ ] Office 365 groups activity #234
- [ ] OneDrive activity #235
- [ ] OneDrive Usage #235
- [ ] SharePoint activity #235
- [ ] SharePoint site usage #235
- [ ] Skype for Business activity #236
- [ ] Skype for Business device usage  #236
- [ ] Skype for Business organizer activity #236
- [ ] Skype for Business participant activity #236
- [ ] Skype for Business peer-to-peer activity #236
- [ ] Yammer activity  #237
- [ ] Yammer device usage  #237
- [ ] Yammer groups activity #237

**Comment by DevLazio** [comment]
How do you plan to implement them ? I will have to fetch and parse some of them for work soon so if you can describe how you would wish to see them implemented I could kill 2 birds with 1 stone.

**Comment by sreeise** [comment]
> 
> 
> How do you plan to implement them ? I will have to fetch and parse some of them for work soon so if you can describe how you would wish to see them implemented I could kill 2 birds with 1 stone.

I believe some, if not most, are already done regarding these. This ticket is somewhat older, and was added before we started parsing and using OpenApi files for Microsoft graph in order to implement the api methods. I try to go through and update or close tickets like these. I need to check what is currently included from this list so here in the next few days ill get back to you on that. Is there a specific graph Api your looking to use?

**Comment by DevLazio** [comment]
I don't have the full list of reports I'll need because specs are still being written but I already now I'll need 
- getOffice365ActivationsUserDetail
- getOffice365ActiveUserDetail  (both by period and date )
- most of the applications activity reports (Teams, OneDrive, Outlook, ...) by period and date

**Comment by sreeise** [comment]
> I don't have the full list of reports I'll need because specs are still being written but I already now I'll need
> 
>     * getOffice365ActivationsUserDetail
> 
>     * getOffice365ActiveUserDetail  (both by period and date )
> 
>     * most of the applications activity reports (Teams, OneDrive, Outlook, ...) by period and date

You can get activity reports for Teams and OneDrive already, and use OData queries to get by date. I will need to get back to you on the others.

**Comment by sreeise** [comment]
So the Reports API has all that as shown in the docs. I have a branch that will include implementation for the reports API and get back to you hopefully within a few weeks on using it.

**Comment by DevLazio** [comment]
Sounds great !
Do tell if you need help for the manual parts or for testing.

**Comment by sreeise** [comment]
@DevLazio Hey, ive been working on getting these done but I have run into a few snags. What I didnt realize at first is that most of these Report API requests are for getting CSV files of the content, and that makes it somewhat difficult to generate given that there has to be some indication in the OpenAPI file that says these reqeusts are going to be a file that you can download after a redirect. 

Getting the response in a file isn't an issue and there are other APIs that are covered that also do redirects to a download page. But the way requests like this work are different and so the code that is generated for downloadable files uses a different macro. 

I am currently going through the OpenAPI config for the Graph API's to see if I can find something that can be used to identify these types of requests. It may be possible that I can use the 302 code that the response comes back with but that is really only meant to say that the request is a redirect with the header having a location field of URL to redirect to, not that its going to be a downloadable file on the other end. Ill keep you updated on what I can find and decide to do.

**Comment by DevLazio** [comment]
Thanks for the update. 

On the online doc they are noted with a Return Type of "Stream" but I can't find that in the OpenApi config.
![image](https://user-images.githubusercontent.com/2280811/141769606-8b7c211c-29b4-4e92-960d-8c98f083cd04.png)
Although it looks like every request that returns a "#/components/schemas/microsoft.graph.report" schema is a CSV file, but that doesn't mean that the opposite is true.
![image](https://user-images.githubusercontent.com/2280811/141770871-25486407-69cb-45c2-a22e-c49f612b020c.png)

The "Content-Disposition" header of the 2nd response might be a better indication if you're going that way.

**Comment by sreeise** [comment]
Yeah, I found the reports schema and one other components specific part in the OpenAPI config will at least work for the reports API. Working out a few issues now on it and possibly after the holidays is when it looks like ill be able to finish it.

**Comment by sreeise** [comment]
@DevLazio If you want, you can start testing things out on [this branch](https://github.com/sreeise/graph-rs/tree/openapi).

I still need to clean up a few things, such as some of the method names. Microsoft, for whatever reason, puts the same name for two different requests and so they put random characters on the end of their names in the OpenAPI config instead of just naming them according to what they do specifically. 

For instance, you can get a report of active user counts by date or by period. For getting it by date, the path is `/reports/getOffice365ActiveUserDetail(date={{id}})` and the method name they gave is `get_office_365_active_user_detail_d_389`. For getting it by period the path is `/reports/getOffice365ActiveUserDetail(period='{{id}}')` and the method name they gave is `get_office_365_active_user_detail_6_8ad`.

**Comment by DevLazio** [comment]
Awesome ! Thanks a lot, I'll start testing right away. 

That's some weird naming scheme indeed.

**Comment by DevLazio** [comment]
Okay, I think I found why the program hangs [here](https://github.com/sreeise/graph-rs/blob/openapi/graph-http/src/download/client.rs#L323) : is_overwrite_existing_file is trying to acquire the mutex lock but can't because it has already been acquired [here](https://github.com/sreeise/graph-rs/blob/openapi/graph-http/src/download/client.rs#L267).

Adding `drop(request);` just above seems to resolve the issue (we get an error if overwrite_existing_file has not been set, file is overwritten otherwise)
![image](https://user-images.githubusercontent.com/2280811/144035881-6f4513e5-c0f3-462f-9a81-35970902a057.png)

**Comment by sreeise** [comment]
@DevLazio Worked on the issue with getting file content a bit more. I updated my branch so that you can now choose whether you want to get the content as text, bytes, or to download the file.

**Comment by DevLazio** [comment]
Just tested and it's exactly what I needed, thank you!

As a side note, I noticed that errors returned are not Send + Sync anymore. I used to return 
`std::result::Result<T, Box<dyn std::error::Error + Send + Sync>>`
from functions using graph_rs and I now have to return 
`std::result::Result<T, Box<dyn std::error::Error>>`.
I don't need the Send + Sync traits because I use the async version, just wanted you to know if that was not intentional.

**Comment by sreeise** [comment]
> Just tested and it's exactly what I needed, thank you!
> 
> As a side note, I noticed that errors returned are not Send + Sync anymore. I used to return `std::result::Result<T, Box<dyn std::error::Error + Send + Sync>>` from functions using graph_rs and I now have to return `std::result::Result<T, Box<dyn std::error::Error>>`. I don't need the Send + Sync traits because I use the async version, just wanted you to know if that was not intentional.

hmm, weird. I don't remember removing anything that was intentionally labeled with `send + sync` but there could also be several reasons for that happening. Im honestly not sure on this yet.

**Comment by sreeise** [comment]
Going to go ahead and close out this ticket.
